### PR TITLE
Selective Model Extension (with specification text)

### PR DESCRIPTION
**Pull request by HansOlsson**
Just trying to get MCP/0032 specification ready.

Whether this should replace MCP/0032 or be merged directly is a latter issue.
First is this ok-ish @christoff-buerger ?

### CODE REVIEW

File: chapters/classes.tex [hunk 1]
```diff
@@ -640,8 +640,8 @@ \section{Class Declarations}\label{class-declarations}
     algorithm-section
   }
   [ external [ language-specification ]
-  [ external-function-call ] [ annotation ] ";" ]
-  [ annotation ";" ]
+  [ external-function-call ] [ annotation-clause ] ";" ]
```
Comment (qlambert-pro) [inline chapters/classes.tex:643]: These modifications although necessary don't seem to be related to this PR

File: chapters/functions.tex [hunk 2]
```diff
@@ -1933,8 +1933,8 @@ \section{External Function Interface}\label{external-function-interface}
   [ protected { component-clause ";" } ]
 external [ language-specification ]
   [ external-function-call ]
-    [annotation ] ";"
-  [ annotation ";" ]
+    [ annotation-clause ] ";"
+  [ annotation-clause ";" ]
```
Comment (qlambert-pro) [inline chapters/functions.tex:1937]: These modifications although necessary don't seem to be related to this PR

**Review by qlambert-pro — COMMENTED** [review]
This seems good to me.

**Comment by HansOlsson** [comment]
> It seems like the intent is that only components and connections are allowed to be deselected, but I'm not really seeing any rule that forbid other elements (like classes) from being deselected

Yes. That seems like a minor issue that should be clarified.

**Comment by christoff-buerger** [comment]
> It seems like the intent is that only components and connections are allowed to be deselected, but I'm not really seeing any rule that forbid other elements (like classes) from being deselected. The semantic rules list components and connections, but that list could be interpreted as just extra rules for those elements.

Only component and connections can be deselected.
As you say yourself, these are the ones listed. The list is exhaustive and normative; it is not non-normative text (everything non-normative is markes as example).

**Review by gkurzbach — APPROVED** [review]
The specification looks good. Last week I did an implementation in SimulationX. During this time there appeared no problems. It turned out that it is really not difficult and a complete implementation can be done in about 2 working days.

**Review by eshmoylova — CHANGES_REQUESTED** [review]
Looks good over all. Just a minor wording comment.

**Comment by HansOlsson** [comment]
Collection of reviews:

- Gerd Kurzbach "The specification looks good. Last week I did an implementation in SimulationX. During this time there appeared no problems. It turned out that it is really not difficult and a complete implementation can be done in about 2 working days." https://github.com/modelica/ModelicaSpecification/pull/3166#pullrequestreview-1097291110_
- Per Öst "It seems like the intent is that only components and connections are allowed to be deselected, but I'm not really seeing any rule that forbid other elements (like classes) from being deselected. The semantic rules list components and connections, but that list could be interpreted as just extra rules for those elements. 
But other than that minor issue it looks good to me."  https://github.com/modelica/ModelicaSpecification/issues/3166#issuecomment-1158866698_
- Quentin Lambert "This seems good to me.; https://github.com/modelica/ModelicaSpecification/pull/3166#pullrequestreview-966061900_

Pending confirmation:
- Elena Shmoylova (pending review): "Looks good over all. Just a minor wording comment."

Will move forward as soon as that is resolved.

**Review by qlambert-pro — APPROVED** [review]

File: chapters/inheritance.tex [hunk 3]
```diff
@@ -1041,3 +1043,131 @@ \subsection{Annotations for Redeclaration and Modification}\label{annotation-cho
...
+\begin{example}
+The syntactic equivalence of connection deselection ensures that connect-statements in for-loops can be deselected:
+\begin{lstlisting}[language=modelica]
+model B
+  $\ldots$
+equation
+  if b then
+    for i in 2:10 loop
+      connect( // Let's add some whitespace for fun.
+        a[i],
```
Comment (dietmarw) [inline chapters/inheritance.tex:1158]: When looking through the new specification text I was thrown off by this and the two white space comments. What is the significance of these? Removing those seems much less confusing to the reader.

Comment (henrikt-ma) [reply chapters/inheritance.tex:1158]: The point is that the equations are not compared for textual equality, implying that white-space and comments will not destroy equivalence.

Comment (dietmarw) [reply chapters/inheritance.tex:1158]: OK, in that case, I would probably remove " for fun" which in this specification gives a non-professional impression. I first thought it was some kind of intermediate test that was forgotten to clean up.

Comment (henrikt-ma) [reply chapters/inheritance.tex:1158]: OK, let's restart this with a conversation attached to the line to be changed.  Resolving.

File: chapters/inheritance.tex [hunk 4]
```diff
@@ -1041,3 +1043,131 @@ \subsection{Annotations for Redeclaration and Modification}\label{annotation-cho
...
+
+\begin{example}
+The syntactic equivalence of connection deselection ensures that connect-statements in for-loops can be deselected:
+\begin{lstlisting}[language=modelica]
+model B
+  $\ldots$
+equation
+  if b then
+    for i in 2:10 loop
+      connect( // Let's add some whitespace for fun.
```
Comment (henrikt-ma) [inline chapters/inheritance.tex:1157]: Following up on conversation with @dietmarw:
```suggestion
      connect( // A comment does not impact equation equivalence.
```

Comment (dietmarw) [reply chapters/inheritance.tex:1157]: Yes much better!

Comment (HansOlsson) [reply chapters/inheritance.tex:1157]: Yes, the idea is that it is about syntactic equivalence, we could even write 'the space and this comment...'

Comment (henrikt-ma) [reply chapters/inheritance.tex:1157]: Sure, but I think that would make the line too long (84 characters).  Hence, I think it's better to not mention the space.

Comment (HansOlsson) [reply chapters/inheritance.tex:1157]: Ok, makes sense.

File: mls.bib [hunk 5]
```diff
@@ -82,3 +82,15 @@ @Manual{Pouzet2006LucidSynchrone30
...
+  author = {Bürger, Christoff},
+  title = {Modelica language extensions for practical non-monotonic modelling: on the need for \emph{selective} model extension},
+  year = {2019},
+  month = mar,
+  day = {4--6},
+  booktitle = {Proceedings of the 13\textsuperscript{th} International Modelica Conference},
+  pages = {277--288},
+  address = {Regensburg, Germany},
+  url = {https://doi.org/10.3384/ecp19157277},
+}
```
Comment (HansOlsson) [inline mls.bib:96]:
```suggestion
}

```
Add missing end-line at end.

Comment (HansOlsson) [reply mls.bib:96]: I think this works. Having a trailing empty line seems preferable.

**Comment by HansOlsson** [comment]
(Will check that the line didn't get too long and that it built.)
